### ISSUE TITLE
Arguments of pulp_nn_conv_Co_parallel

### ISSUE THREAD

**Issue by MiguelCosta94** [pulp-platform/pulp-nn#3]
Hi,

I am trying to implement a CNN using the master version of PULP-NN. However, I am experience some difficulties in understanding some parameters of the function "pulp_nn_conv_Co_parallel". Can you provide more insights about the meaning of the following parameters?

- out_mult
- k
- lambda
- memory_chan

**Comment by NBruschi** [comment]
Hello Miguel,

sure! PULP-NN kernels intrinsically include quantization, activation and normalization steps of the outputs, which are accumulated on int32_t variables. For these reason, they embed two different types of output blocks, which you can pick using flag_relu and flag_batch_norm arguments:

- Quantization + Normalization (flag_relu=1, flag_batch_norm=1)
- Quantization (flag_relu=X, flag_batch_norm=0)

The first is quite useful if these kernels are used in the Convolutional Neural Network (CNN) in order to normalize (and activate with a ReLu) the output results, while the last is udeful to restore the precision of the output results to uint8_t, using a either mult+shift+clipping procedure or shift+clipping procedure (flag_relu=1, flag_relu=0 respectively).
In particular, kappa and lambda are pointers to batch normalization parameters that, in this scheme, are ch_out each. out_mult is instead a single parameter for the quantization.
Of course, the choice depends from the application and from the input, because, accumulating the results, they can become bigger soon and not be represented in an uint8_t variable.

memory_chan is a different parameter, which can be set to NULL if is not necessary. It is useful if you want to program the DMA of PULP cluster, in order to pass to the convolution programming information.

**Comment by MiguelCosta94** [comment]
Hello @NBruschi,

Thanks for your reply. Can you provide more insights on how to define kappa and lambda parameters? I am trying to port the implementation of a simple CNN, originally implemented using PULP-NN but in the "paper version" branch, to the new version of PULP-NN, in the "master" branch. I am experience a drop in accuracy from 94% to less than 9%. Both versions follow the Height-Width-Channel (HWC) layout right?

**Comment by NBruschi** [comment]
Remember that the "paper version" takes as input and produces output int8_t, while this version uint8_t. This extends the range of the values avoiding the bit for the sign. For conv, pointwise, linear and depthwise, the weights still int8_t.
kappa and lambda are in particular a scale factor and lambda an additive factor that typically represents the mean and the variance. Batch normalization is a normalization (https://en.wikipedia.org/wiki/Batch_normalization), so your CNN could have it or not. If you want to experiment with batch norm you should process your input and evaluate these two parameters. If you want to exclude this block like the "paper version" does, you should set both flags to zero, using the shift+clip quantization.
Regarding the HWC layout, every kernel takes input and produces output in this fashion excluding the depthwise kernel which, to speed-up the computation, takes input in CHW fashion and produces output in HWC to fit the data layout of the next layer.

Do you know this tool https://github.com/pulp-platform/dory? It is a tool to support the deployment of some CNNs on PULP, stressing a lot PULP-NN. If you would take inspiration for your work, we would appreciate it a lot.

**Comment by MiguelCosta94** [comment]
Hi  @NBruschi,

Thank you so much for your reply. I was not spotting that. The difference in i/o format between the "paper version" and the "master version" was the main source of error in my program. 
I was using this scheme ( https://github.com/ARM-software/ML-KWS-for-MCU/blob/master/Deployment/Quant_guide.md ) to quantize my original CNN, trained in Tensorflow 2. As my original CNN has negative activations between layers, I guess I will need to use another quantization scheme. I will take a look at DORY to try to find a solution.
